src/mod_webdav.c

```diff
@@ -1750,33 +1758,51 @@ static handler_t mod_webdav_put(server *srv, connection *con, plugin_data *p, ha
 		 * Example: Content-Range: bytes 100-1037/1038 */
 
 		if (NULL != (b = http_header_request_get(con, HTTP_HEADER_OTHER, CONST_STR_LEN("Content-Range")))) {
-			const char *num = b->ptr;
-			off_t offset;
-			char *err = NULL;
+			char *num = b->ptr;
+			off_t request_length = -1;
+			off_t offset, complete_length = -1;
+			char *err = num;  /* err == NULL -> parse success */
+
+			/* parse range */
+			do {
+				if (0 != strncmp(num, "bytes ", 6)) break;
+				num += 6;
+
+				/* skip WS */
+				while (*num == ' ' || *num == '\t') num++;
+
+				offset = strtoll(num, &err, 10);
+				if (num == err || *err != '-' || offset < 0) break;
+				num = err;
+
+				num++;
+				if (*num != '\0') {
+					off_t end;
+					end = strtoll(num, &err, 10);
+					if (num == err || *err != '/' || end < offset) break;
+					request_length = end - offset + 1;
+					num = err;
+
+					num++;
+					if (*num != '*') {
+						complete_length = strtoll(num, &err, 10);
+						if (num == err || (complete_length > 0 && complete_length <= end)) break;
+					}
+				}
+				err = NULL;
+			} while (0);
 
-			if (0 != strncmp(num, "bytes ", 6)) {
+			if (NULL != err) {
+				/* parse error */
 				con->http_status = 501; /* not implemented */
 
 				return HANDLER_FINISHED;
 			}
 
-			/* we only support <num>- ... */
-
-			num += 6;
-
-			/* skip WS */
-			while (*num == ' ' || *num == '\t') num++;
-
-			if (*num == '\0') {
-				con->http_status = 501; /* not implemented */
-
-				return HANDLER_FINISHED;
-			}
-
-			offset = strtoll(num, &err, 10);
-
-			if (*err != '-' || offset < 0) {
-				con->http_status = 501; /* not implemented */
+			/* TODO: if uploaded content length > request_length, the remain should be
```

Comment:
Not a TODO.  con->request_content_queue contains exactly the request body contents.  mod_webdav waits until everything is uploaded and stores the request body in temporary files on local disk if the request body is large.